### PR TITLE
Added options call for cloud networks

### PR DESCRIPTION
**Pull request by GilbertCherrie**
Depends on: #1060

Added an options call for cloud networks:

Calling example with no cloud_network_id:
OPTIONS "http://admin:smartvm@localhost:3000/api/cloud_networks?ems_id=8"  
`{
  "attributes": [
    "cidr",
    "cloud_tenant_id",
    "description",
    "ems_id",
    "ems_ref",
    "enabled",
    "external_facing",
    "extra_attributes",
    "id",
    "name",
    "orchestration_stack_id",
    "provider_network_type",
    "provider_physical_network",
    "provider_segmentation_id",
    "resource_group_id",
    "shared",
    "status",
    "type",
    "vlan_transparent"
  ],
  "virtual_attributes": [
    "href_slug",
    "maximum_transmission_unit",
    "port_security_enabled",
    "qos_policy_id",
    "region_description",
    "region_number",
    "total_vms"
  ],
  "relationships": [
    "cloud_subnets",
    "cloud_tenant",
    "custom_action_buttons",
    "custom_actions",
    "custom_button_events",
    "custom_button_sets",
    "custom_buttons",
    "ext_management_system",
    "floating_ips",
    "network_ports",
    "network_routers",
    "orchestration_stack",
    "private_networks",
    "public_network_routers",
    "public_network_vms",
    "public_networks",
    "resource_group",
    "security_groups",
    "taggings",
    "tags",
    "vms"
  ],
  "subcollections": [
    "tags"
  ],
  "data": {
    "form_schema": {
      "fields": [
        {
          "component": "sub-form",
          "title": "Placement",
          "id": "placement",
          "name": "placement",
          "fields": [
            {
              "component": "select",
              "id": "cloud_tenant",
              "name": "cloud_tenant",
              "key": "id-8",
              "label": "Cloud Tenant",
              "placeholder": "<Choose>",
              "validateOnMount": true,
              "validate": [
                {
                  "type": "required",
                  "message": "Required"
                }
              ],
              "isDisabled": false,
              "options": [
                {
                  "label": "cloud-user-demo",
                  "value": "1"
                },
                {
                  "label": "admin",
                  "value": "2"
                },
                {
                  "label": "openshift_demo",
                  "value": "3"
                },
                {
                  "label": "testtel",
                  "value": "4"
                },
                {
                  "label": "cloud-southeast",
                  "value": "5"
                },
                {
                  "label": "Loic Tenant",
                  "value": "6"
                },
                {
                  "label": "cloudwest",
                  "value": "7"
                },
                {
                  "label": "Massachusetts",
                  "value": "8"
                },
                {
                  "label": "testetot",
                  "value": "10"
                },
                {
                  "label": "test-ivy",
                  "value": "13"
                },
                {
                  "label": "test-ivr",
                  "value": "12"
                }
              ],
              "includeEmpty": true,
              "clearOnUnmount": true
            }
          ]
        },
        {
          "component": "sub-form",
          "title": "Network Provider Information",
          "id": "provider-information",
          "name": "provider-information",
          "condition": {
            "when": "cloud_tenant",
            "isNotEmpty": true
          },
          "fields": [
            {
              "component": "select",
              "id": "provider_network_type",
              "name": "provider_network_type",
              "label": "Provider Network Type",
              "placeholder": "Nothing selected",
              "options": [
                {
                  "label": "None"
                },
                {
                  "label": "Local",
                  "value": "local"
                },
                {
                  "label": "Flat",
                  "value": "flat"
                },
                {
                  "label": "GRE",
                  "value": "gre"
                },
                {
                  "label": "GENEVE",
                  "value": "geneve"
                },
                {
                  "label": "VLAN",
                  "value": "vlan"
                },
                {
                  "label": "VXLAN",
                  "value": "vxlan"
                }
              ],
              "isDisabled": false
            },
            {
              "component": "sub-form",
              "id": "subform-1",
              "name": "subform-1",
              "condition": {
                "when": "provider_network_type",
                "is": "flat"
              },
              "fields": [
                {
                  "component": "text-field",
                  "label": "Physical Network",
                  "maxLength": 128,
                  "id": "provider_physical_network",
                  "name": "provider_physical_network",
                  "isDisabled": false,
                  "validateOnMount": true,
                  "clearOnUnmount": true,
                  "validate": [
                    {
                      "type": "required",
                      "message": "Required"
                    }
                  ]
                }
              ]
            },
            {
              "component": "sub-form",
              "id": "subform-2",
              "name": "subform-2",
              "condition": {
                "when": "provider_network_type",
                "is": "gre"
              },
              "fields": [
                {
                  "component": "text-field",
                  "label": "Segmentation ID",
                  "maxLength": 128,
                  "id": "provider_segmentation_id",
                  "name": "provider_segmentation_id",
                  "isDisabled": false,
                  "validateOnMount": true,
                  "clearOnUnmount": true,
                  "validate": [
                    {
                      "type": "required",
                      "message": "Required"
                    }
                  ]
                }
              ]
            },
            {
              "component": "sub-form",
              "id": "subform-3",
              "name": "subform-3",
              "condition": {
                "when": "provider_network_type",
                "is": "vlan"
              },
              "fields": [
                {
                  "component": "text-field",
                  "label": "Physical Network",
                  "maxLength": 128,
                  "id": "provider_physical_network",
                  "name": "provider_physical_network",
                  "isDisabled": false,
                  "validateOnMount": true,
                  "clearOnUnmount": true,
                  "validate": [
                    {
                      "type": "required",
                      "message": "Required"
                    }
                  ]
                },
                {
                  "component": "text-field",
                  "label": "Segmentation ID",
                  "maxLength": 128,
                  "id": "provider_segmentation_id",
                  "name": "provider_segmentation_id",
                  "isDisabled": false,
                  "validateOnMount": true,
                  "clearOnUnmount": true,
                  "validate": [
                    {
                      "type": "required",
                      "message": "Required"
                    }
                  ]
                }
              ]
            },
            {
              "component": "sub-form",
              "id": "subform-4",
              "name": "subform-4",
              "condition": {
                "when": "provider_network_type",
                "is": "vxlan"
              },
              "fields": [
                {
                  "component": "text-field",
                  "label": "Segmentation ID",
                  "maxLength": 128,
                  "id": "provider_segmentation_id",
                  "name": "provider_segmentation_id",
                  "isDisabled": false,
                  "clearOnUnmount": true
                }
              ]
            }
          ]
        },
        {
          "component": "sub-form",
          "title": "Network Information",
          "id": "network-information",
          "name": "network-information",
          "fields": [
            {
              "component": "text-field",
              "id": "name",
              "name": "name",
              "validateOnMount": true,
              "label": "Network Name",
              "validate": [
                {
                  "type": "required",
                  "message": "Required"
                }
              ]
            }
          ]
        },
        {
          "component": "switch",
          "id": "cloud_network_external_facing",
          "name": "external_facing",
          "label": "External Router",
          "onText": "Yes",
          "offText": "No"
        },
        {
          "component": "switch",
          "id": "cloud_network_enabled",
          "name": "enabled",
          "label": "Adminstrative State",
          "onText": "Up",
          "offText": "Down"
        },
        {
          "component": "switch",
          "id": "cloud_network_shared",
          "name": "shared",
          "label": "Shared",
          "onText": "Yes",
          "offText": "No"
        }
      ]
    }
  }
}`

Can also call with a cloud_network_id in order to return only fields that are allowed to be edited while the rest are disabled.
OPTIONS "http://admin:smartvm@localhost:3000/api/cloud_networks/${cloud_network_id}

@miq-bot add_reviewer @Fryguy 
@miq-bot add-label enhancement
@miq-bot add-label dependencies

### CODE REVIEW

**Comment by agrare** [comment]
We need to standardize how we want to handle provider defined DDF params (`params_for_create` with an instance_id vs. standalone `params_for_update`) and how we access those from the API (`OPTIONS /collection?instance_id=:id` VS `OPTIONS /collection/:id`)

We happen to have two pending PRs doing this two different ways: https://github.com/ManageIQ/manageiq-api/pull/1056

cc @Fryguy

**Comment by agrare** [comment]
If we're getting `OPTIONS` for a record, prefer to do `OPTIONS /api/cloud_networks/:id` rather than `OPTIONS /api/cloud_networks?cloud_network_id=:id`

Splitting up `CloudNetwork#params_for_edit` from `CloudNetwork.params_for_create` is also more obvious, calling params_for_create with an optional instance ID is confusing.  The EMS params_for_create gets away with this because nothing is not able to be edited but that doesn't hold true for all of these types.

**Review by NickLaMuro — CHANGES_REQUESTED** [review]
@GilbertCherrie Can you please create a global `.gitignore` file:

https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer

and add the following files to it:

- `.yarn-integrity`
- `.DS_Store`
- `yarn.lock`

Also, please remove said files from your commits.

**Comment by GilbertCherrie** [comment]
@agrare The params for edit would need the ems_id and cloud_network_id passed to it so should I call it as           `API.options(`/api/cloud_networks/${ems_id}?cloud_network_id=${cloudNetworkId}`)` and for create I would just do
`API.options(`/api/cloud_networks/${ems_id}`?

**Comment by agrare** [comment]
@GilbertCherrie params_for_edit is going to be an instance method so you don't need to give it either of those as params because they are available as part of the cloud_network instance.

E.g. if you call `OPTIONS /api/cloud_networks/${cloudNetworkId}` the EMS is available to the object as `ext_management_system` and cloudNetworkId is just `id`

**Comment by GilbertCherrie** [comment]
> @GilbertCherrie params_for_edit is going to be an instance method so you don't need to give it either of those as params because they are available as part of the cloud_network instance.
> 
> E.g. if you call `OPTIONS /api/cloud_networks/${cloudNetworkId}` the EMS is available to the object as `ext_management_system` and cloudNetworkId is just `id`

So I don't need to pass anything to params_for_edit and it would have access to ems and the id?

**Comment by agrare** [comment]
> So I don't need to pass anything to params_for_edit and it would have access to ems and the id?

You're loading the record [[here]](https://github.com/ManageIQ/manageiq-api/pull/1058/files#diff-4c387b2b1fa70c078e35f28c11fa2dea10961698eb5db17d745d3c937e31a09bR27), that means that when you call the instance method `#params_for_edit` you are in the context of the instance aka you have access to its attributes.  The ems_id is one attribute which is equivalent to the `${ems_id}` that you were passing in.

**Comment by GilbertCherrie** [comment]
> > So I don't need to pass anything to params_for_edit and it would have access to ems and the id?
> 
> You're loading the record [[here]](https://github.com/ManageIQ/manageiq-api/pull/1058/files#diff-4c387b2b1fa70c078e35f28c11fa2dea10961698eb5db17d745d3c937e31a09bR27), that means that when you call the instance method `#params_for_edit` you are in the context of the instance aka you have access to its attributes. The ems_id is one attribute which is equivalent to the `${ems_id}` that you were passing in.

I think I understand, but one last point of confusion for me is how does the api code differentiate between ems_id and id if the API call is simply done as OPTIONS /api/cloud_networks/${ems_id}. The {ems_id} would just be a number so how would it know if this is an ems_id or cloud_network_id?

**Comment by agrare** [comment]
> api code differentiate between ems_id and id if the API call is simply done as OPTIONS /api/cloud_networks/${ems_id}.

When you call `OPTIONS /api/cloud_networks/:id` the API just sees `/api/cloud_networks/1` for example.  The convention is that number is the id of the record in the (in this case) cloud_networks table.  That is the equivalent of doing `CloudNetwork.find(1)` in a rails console (which actually is what you are doing here: https://github.com/ManageIQ/manageiq-api/pull/1058/files#diff-4c387b2b1fa70c078e35f28c11fa2dea10961698eb5db17d745d3c937e31a09bR27

Remember you aren't passing ems_id to this call, you're passing cloudNetworkId.

If you did `GET /api/cloud_volumes/12` that would be equivalent to `CloudVolume.find(12)`.

The API doesn't know anything about the ems_id.  You can get the ems_id by virtue of the fact that you're already working with a single record, and that record has an ems_id.

File: app/controllers/api/cloud_networks_controller.rb
Comment (agrare) [inline app/controllers/api/cloud_networks_controller.rb:17]: Should use the `resource_search` method here: `resource_search(params["ems_id"], :ext_management_systems, ExtManagementSystem)`

File: app/controllers/api/cloud_networks_controller.rb
Comment (agrare) [inline app/controllers/api/cloud_networks_controller.rb:25]: Same thing here, `resource_search(params["id"], :cloud_networks, CloudNetwork)`

File: spec/requests/cloud_networks_spec.rb
Comment (agrare) [inline spec/requests/cloud_networks_spec.rb:114]: Should also add an `'OPTIONS /api/cloud_networks/:id` spec here

File: app/controllers/api/cloud_networks_controller.rb
Comment (GilbertCherrie) [reply app/controllers/api/cloud_networks_controller.rb:17]: `ems = resource_search(params["ems_id"], :ems, ExtManagementSystem)`

Is this correct?

File: spec/requests/cloud_networks_spec.rb
Comment (GilbertCherrie) [reply spec/requests/cloud_networks_spec.rb:114]: @agrare for this test case I currently have:
`
allow(provider.class::CloudNetwork).to receive(:params_for_edit).and_return('foo')

options(api_cloud_networks_url, :params => {})
options("#{api_cloud_networks_url}/#{provider.id}")

expect(response.parsed_body['data']['form_schema']).to eq('foo')
expect(response).to have_http_status(:ok)
`
Not sure if I am setting this up wrong but is there any way I can get the cloud_network_id from provider because the way I have it now the test case is failing due to the options call not returning anything.

**Comment by GilbertCherrie** [comment]
I believe failed tests are due to this PR: https://github.com/ManageIQ/manageiq-api/pull/1060 not being merged yet. Test cases pass locally with this PR included and should pass once it is merged.

**Comment by agrare** [comment]
@GilbertCherrie just throw `Depends on: #1060` in the description of this PR

**Comment by GilbertCherrie** [comment]
@agrare @Fryguy I saw this PR: https://github.com/ManageIQ/manageiq-api/pull/1060 get merged. If any of you have time would you be able to review and merge this PR as well?

**Comment by agrare** [comment]
Seeing as this is a new API call, @Fryguy or @bdunne can you also review?

File: app/controllers/api/cloud_networks_controller.rb
Comment (agrare) [inline app/controllers/api/cloud_networks_controller.rb:18]: I know this was just copied&pasted from elsewhere so not going to block this PR for it, but in general checking `supports?(:cloud_network)` on the manager would be better than check class names existing.

None of these features exist right now so I think it'd be better to add these and update the API in one PR set

File: app/controllers/api/cloud_networks_controller.rb
Comment (agrare) [inline app/controllers/api/cloud_networks_controller.rb:21]: We should be checking for `klass.supports?(:create)` instead of checking for `respond_to?(:params_for_create)` here

File: app/controllers/api/cloud_networks_controller.rb
Comment (agrare) [inline app/controllers/api/cloud_networks_controller.rb:20]: `CloudNetwork.class_by_ems(ems)` is the preferred way of getting the class name

File: app/controllers/api/cloud_networks_controller.rb
Comment (agrare) [inline app/controllers/api/cloud_networks_controller.rb:23]: You already have the `klass` above so you can do `klass.params_for_create(ems)` here

**Review by agrare — COMMENTED** [review]
Couple of minor things here

File: app/controllers/api/cloud_networks_controller.rb
Comment (GilbertCherrie) [reply app/controllers/api/cloud_networks_controller.rb:20]: Which line is this referring to?

Comment (agrare) [reply app/controllers/api/cloud_networks_controller.rb:20]: this one

```suggestion
        klass = CloudNetwork.class_by_ems(ems)
```

**Comment by agrare** [comment]
@GilbertCherrie can you update the description of your UI PR to indicate that it depends on this one?

File: app/controllers/api/cloud_networks_controller.rb
Comment (agrare) [inline app/controllers/api/cloud_networks_controller.rb:20]: Sorry think I confused you, we don't have `cloud_networks` as a feature that we are tracking so you can't do `supports?(:cloud_network)` yet.

You can keep the `defined?(ems.class::CloudNetwork)` for now

**Review by agrare — APPROVED** [review]

File: app/controllers/api/cloud_networks_controller.rb
Comment (Fryguy) [inline app/controllers/api/cloud_networks_controller.rb:16]: I think this might be cleaner as follows:

```ruby
def options
  if params.key?("ems_id")
    options_by_ems_id
  elsif params.key?("id")
    options_by_id
  else
    super
  end
end
```

Then break out these other 2 sections into their own private methods `options_by_ems_id` and `options_by_id`. 

WDYT?

Comment (Fryguy) [reply app/controllers/api/cloud_networks_controller.rb:16]: (Maybe not for this PR, especially if this pattern was copy/pasted)

Comment (GilbertCherrie) [reply app/controllers/api/cloud_networks_controller.rb:16]: I think splitting them up is cleaner, I can change it right now if you want it split up

File: app/controllers/api/cloud_networks_controller.rb
Comment (Fryguy) [inline app/controllers/api/cloud_networks_controller.rb:17]: I can never remember, but does `resource_search` handle the RBAC?

File: app/controllers/api/cloud_networks_controller.rb
Comment (Fryguy) [reply app/controllers/api/cloud_networks_controller.rb:16]: Go for it 😊

File: app/controllers/api/cloud_networks_controller.rb
Comment (agrare) [reply app/controllers/api/cloud_networks_controller.rb:17]: Yes

**Comment by Fryguy** [comment]
Overall LGTM - just have one question on the RBAC.

@NickLaMuro we also need your re-review.

File: app/controllers/api/cloud_networks_controller.rb
Comment (Fryguy) [reply app/controllers/api/cloud_networks_controller.rb:17]: If the user cannot see it, do we get a nil back or does it instantly 404?

Comment (agrare) [reply app/controllers/api/cloud_networks_controller.rb:17]: That I don't know

File: app/controllers/api/cloud_networks_controller.rb
Comment (Fryguy) [inline app/controllers/api/cloud_networks_controller.rb:13]: Can you make `options_by_ems_id` and `options_by_id` private? (Either prefix with `private` or put them in a `private` section at the bottom).

File: spec/requests/cloud_networks_spec.rb
Comment (Fryguy) [inline spec/requests/cloud_networks_spec.rb:109]: This feels like it should be 2 separate specs as opposed to calling one after the other.  I don't think the expects on lines 111-112 will pick up on anything set in the first call (or put another way, if you delete line 108, this test will still pass, so it's not actually being tested)

Comment (GilbertCherrie) [reply spec/requests/cloud_networks_spec.rb:109]: I just tested it as 2 different specs, the spec with just line 108 failed. Do I keep it as 2 specs or leave it as 1 spec but remove line 108?

Comment (Fryguy) [reply spec/requests/cloud_networks_spec.rb:109]: I would keep it as 2 specs, and then we have to figure out why the first line fails.  If this was a copy past, we'll have to figure out why this isn't working.

Offhand, I'm wondering if it's a String vs Symbol thing (i.e. `:ems_id` vs `"ems_id"`), since that's the key you are looking for, but I thought `params` handled that.

Comment (GilbertCherrie) [reply spec/requests/cloud_networks_spec.rb:109]: @Fryguy I just checked the cloud_subnets_spec.rb which is where this line initially came from, 
`    options(api_cloud_subnets_url, :params => {:ems_id => provider.id})
      options("#{api_cloud_subnets_url}?ems_id=#{provider.id}")`
After repeating the same process of removing the second options call and only testing the first the test case fails, so I am wondering if possibly the options call is being used incorrectly. I also checked the string vs symbol issue and it did not fix the problem.

Comment (Fryguy) [reply spec/requests/cloud_networks_spec.rb:109]: Interesting.  @agrare maybe you have some idea?

Comment (agrare) [reply spec/requests/cloud_networks_spec.rb:109]: I have no idea why they call `options` twice but only check the response once

Comment (agrare) [reply spec/requests/cloud_networks_spec.rb:109]: welp this is interesting

```
     def options
+      raise
       if params.key?("ems_id")
         options_by_ems_id
       elsif params.key?("id")
diff --git a/spec/requests/cloud_networks_spec.rb b/spec/requests/cloud_networks_spec.rb
index b0633efa..7387ec4f 100644
--- a/spec/requests/cloud_networks_spec.rb
+++ b/spec/requests/cloud_networks_spec.rb
@@ -106,7 +106,7 @@ RSpec.describe 'Cloud Networks API' do
       allow(provider.class::CloudNetwork).to receive(:params_for_create).and_return('foo')
 
       options(api_cloud_networks_url, :params => {:ems_id => provider.id})
-      options("#{api_cloud_networks_url}?ems_id=#{provider.id}")
+      #options("#{api_cloud_networks_url}?ems_id=#{provider.id}")
```

That first one isn't invoking the overridden options call at all?

Comment (agrare) [reply spec/requests/cloud_networks_spec.rb:109]:
```
response.parsed_body
{"error"=>{"kind"=>"bad_request", "message"=>"809: unexpected token at 'ems_id=5000000000028'", "klass"=>"JSON::ParserError", 
```

Comment (GilbertCherrie) [reply spec/requests/cloud_networks_spec.rb:109]: @agrare I have been trying out different ways of calling this and none seem to work:

`options(api_cloud_networks_url, :params => {:ems_id => provider.id})`

This line is returning a 400 error, but when I remove the ems_id and leave params as empty (:params => {}) then it returns a 200 response. So is the problem with the ems_id?

Comment (agrare) [reply spec/requests/cloud_networks_spec.rb:109]: What's wrong with `options("#{api_cloud_networks_url}?ems_id=#{provider.id}")`

Comment (GilbertCherrie) [reply spec/requests/cloud_networks_spec.rb:109]: That one works, but @Fryguy wanted me to split up the 2 options call into separate specs so one calls this as well:
`options(api_cloud_networks_url, :params => {:ems_id => provider.id})`

Comment (agrare) [reply spec/requests/cloud_networks_spec.rb:109]: I'm wondering if that is even valid, I don't know where that came from or if it is expected to work.

Maybe a question for someone deeper into our API than me.

Comment (GilbertCherrie) [reply spec/requests/cloud_networks_spec.rb:109]: I saw it in cloud_subnets_spec.rb so I used that as a starting point for my test cases. When I tried to put it as a separate  test case it failed and then I went back to cloud_subnets_spec.rb and tried the test case with only this options call and saw the same result of the test case failing. So I am wondering if we are either using `options(api_cloud_networks_url, :params => {:ems_id => provider.id})` incorrectly or if it is simply not expected to work.

Comment (agrare) [reply spec/requests/cloud_networks_spec.rb:109]: > So I am wondering if we are either using options(api_cloud_networks_url, :params => {:ems_id => provider.id}) incorrectly or if it is simply not expected to work.

Yes that's exactly what I'm wondering as well

Comment (GilbertCherrie) [reply spec/requests/cloud_networks_spec.rb:109]: @Fryguy would it be possible to leave it as just 1 test case with the working api call:
`options("#{api_cloud_networks_url}?ems_id=#{provider.id}")`?

Comment (Fryguy) [reply spec/requests/cloud_networks_spec.rb:109]: Yeah, for now let's leave as just 1 working test case, and then can you open up a separate GitHub issue for the other one that we can remember to go back and investigate later?

Comment (GilbertCherrie) [reply spec/requests/cloud_networks_spec.rb]: I made the change to the spec file and will open up the issue now.

**Review by Fryguy — APPROVED** [review]

**Review by NickLaMuro — APPROVED** [review]

**Comment by Fryguy** [comment]
Backported to `morphy` in commit b6713cede8856ae07b95eb8017398123b97e8ea7.

```text
commit b6713cede8856ae07b95eb8017398123b97e8ea7
Author: Jason Frey <fryguy9@gmail.com>
Date:   Fri Jul 30 11:36:39 2021 -0400

    Merge pull request #1058 from GilbertCherrie/add_options_call_for_cloud_network
    
    Added options call for cloud networks
    
    (cherry picked from commit f1fa1d853e2bb3432e7eb875a8de6e44d618b6c9)
```

**Pull request merged**
